OpenRA.Mods.Common/Traits/World/PathFinder.cs

```diff
@@ -8,236 +8,237 @@
+
+			return path;
 		}
 
-		public List<CPos> FindUnitPathToRange(CPos src, SubCell srcSub, WPos target, WRange range, Actor self)
+		public List<CPos> FindPath(IPathSearch search)
 		{
-			using (new PerfSample("Pathfinder"))
+			var dbg = world.WorldActor.TraitOrDefault<PathfinderDebugOverlay>();
+			if (dbg != null && dbg.Visible)
+				search.Debug = true;
+
+			List<CPos> path = null;
+
+			while (!search.OpenQueue.Empty)
 			{
-				var mi = self.Info.Traits.Get<MobileInfo>();
-				var targetCell = self.World.Map.CellContaining(target);
-				var rangeSquared = range.Range * range.Range;
-
-				// Correct for SubCell offset
-				target -= self.World.Map.OffsetOfSubCell(srcSub);
-
-				// Select only the tiles that are within range from the requested SubCell
-				// This assumes that the SubCell does not change during the path traversal
-				var tilesInRange = world.Map.FindTilesInCircle(targetCell, range.Range / 1024 + 1)
-					.Where(t => (world.Map.CenterOfCell(t) - target).LengthSquared <= rangeSquared &&
-						mi.CanEnterCell(self.World, self, t));
-
-				// See if there is any cell within range that does not involve a cross-domain request
-				// Really, we only need to check the circle perimeter, but it's not clear that would be a performance win
-				var domainIndex = self.World.WorldActor.TraitOrDefault<DomainIndex>();
-				if (domainIndex != null)
+				var p = search.Expand();
+				if (search.IsTarget(p))
 				{
-					var passable = mi.GetMovementClass(world.TileSet);
-					tilesInRange = new List<CPos>(tilesInRange.Where(t => domainIndex.IsPassable(src, t, (uint)passable)));
-					if (!tilesInRange.Any())
-						return EmptyPath;
+					path = MakePath(search.Graph, p);
+					break;
 				}
+			}
+
+			if (dbg != null && dbg.Visible)
+				dbg.AddLayer(search.Considered, search.MaxCost, search.Owner);
 
-				var path = FindBidiPath(
-					PathSearch.FromPoints(world, mi, self, tilesInRange, src, true),
-					PathSearch.FromPoint(world, mi, self, src, targetCell, true).Reverse());
+			search.Graph.Dispose();
 
+			if (path != null)
 				return path;
-			}
+
+			// no path exists
+			return EmptyPath;
 		}
 
-		public List<CPos> FindPath(PathSearch search)
+		// Searches from both ends toward each other. This is used to prevent blockings in case we find
+		// units in the middle of the path that prevent us to continue.
+		public List<CPos> FindBidiPath(IPathSearch fromSrc, IPathSearch fromDest)
 		{
-			using (new PerfSample("Pathfinder"))
+			List<CPos> path = null;
+
+			var dbg = world.WorldActor.TraitOrDefault<PathfinderDebugOverlay>();
+			if (dbg != null && dbg.Visible)
+			{
+				fromSrc.Debug = true;
+				fromDest.Debug = true;
+			}
+
+			while (!fromSrc.OpenQueue.Empty && !fromDest.OpenQueue.Empty)
 			{
-				using (search)
+				// make some progress on the first search
+				var p = fromSrc.Expand();
+
+				if (fromDest.Graph[p].Status == CellStatus.Closed &&
+					fromDest.Graph[p].CostSoFar < float.PositiveInfinity)
```

Comment:
You've changed `CostSoFar` to an int, so this `float.PositiveInfinity` check should change too.